### ISSUE TITLE
Combine the PacBio HiFi and Ultra-long have worse phase performance

### ISSUE THREAD

**Issue by baozg** [whatshap/whatshap#336]
Hi,

I use the freebayes to call SNPs and InDels in polypoidy genotype from HiFi aligned bam (120x for haploid genome). And also have the Ultra-long reads aligned bam (reads N50 50kb) 60x. But the best performance was using HiFi only. How can I use the Ultra-long information for more contigous phase blocks? 

- HiFi bam for whatshap
```bash
---------------- Chromosome chr01 ----------------
     Variants in VCF:  2439041
        Heterozygous:  2438804 ( 1835527 SNVs)
              Phased:   538109 (  538109 SNVs)
            Unphased:  1898858 (not considered below)
          Singletons:     1837 (not considered below)
              Blocks:    11947

Block sizes (no. of variants)
   Median block size:       16.00 variants
  Average block size:       45.04 variants
       Largest block:      908    variants
      Smallest block:        2    variants

Block lengths (basepairs)
      Sum of lengths: 61984356    bp
 Median block length:     1539.00 bp
Average block length:     5188.28 bp
       Longest block:   132504    bp
      Shortest block:        1    bp
           Block N50:      nan    bp
```

- HiFi + UL ONT

```bash
---------------- Chromosome chr01 ----------------
     Variants in VCF:  2439041
        Heterozygous:  2397651 ( 1794374 SNVs)
              Phased:   494623 (  494623 SNVs)
            Unphased:  1893717 (not considered below)
          Singletons:     9311 (not considered below)
              Blocks:    21083

Block sizes (no. of variants)
   Median block size:        8.00 variants
  Average block size:       23.46 variants
       Largest block:      824    variants
      Smallest block:        2    variants

Block lengths (basepairs)
      Sum of lengths: 63201844    bp
 Median block length:      749.00 bp
Average block length:     2997.76 bp
       Longest block:    97118    bp
      Shortest block:        1    bp
           Block N50:      nan    bp
```

- UL only

```bash

---------------- Chromosome chr01 ----------------
     Variants in VCF:  2439041
        Heterozygous:  2249966 ( 1646689 SNVs)
              Phased:   265943 (  265943 SNVs)
            Unphased:  1901918 (not considered below)
          Singletons:    82105 (not considered below)
              Blocks:    74537

Block sizes (no. of variants)
   Median block size:        3.00 variants
  Average block size:        3.57 variants
       Largest block:      251    variants
      Smallest block:        2    variants

Block lengths (basepairs)
      Sum of lengths: 33566625    bp
 Median block length:      164.00 bp
Average block length:      450.34 bp
       Longest block:    47089    bp
      Shortest block:        1    bp
           Block N50:      nan    bp
```

### Command line
```bash
# minimap2 2.22 for hifi
minimap2 -ax map-hifi haploid.fasta poly.HiFi.fq.gz -t 64|samtools view -@ 64 -Sb -|samtools sort -@ 32 -o poly.hifi.sorted.bam -

samtools addreplacerg -r "@RG\tID:HiFi\tPL:PacBio\tLB:HiFi\tSM:HiFi" -o poly.hifi.addrg.sorted.bam poly.hifi.sorted.bam

freebayes-parallel <(fasta_generate_regions.py haploid.fasta.fai 100000) 64 -f haploid.fasta -p 4 --use-best-n-alleles 4 --skip-coverage 200 poly.hifi.addrg.sorted.bam > poly.hifi.vcf

# Ultra-long 
minimap2 -ax map-ont haploid.fasta poly.ont.fq.gz -t 64|samtools view -@ 64 -Sb -|samtools sort -@ 32 -o poly.ontsorted.bam -

samtools addreplacerg -r "@RG\tID:ONT\tPL:Nanopore\tLB:ONT\tSM:ONT" -o poly.ont.addrg.sorted.bam poly.ont.sorted.bam

# whatshap (v1.1)
## ont and hifi
whatshap polyphase -p 4 -o poly.whatshap.chr1.vcf.gz --threads 64 --verify-genotypes --ignore-read-groups --reference haploid.fasta vcf/chr1.vcf.gz poly.hifi.addrg.sorted.bam poly.ont.addrg.sorted.bam

## hifi only
whatshap polyphase -p 4 -o poly.whatshap.chr1.vcf.gz --threads 64 --verify-genotypes --ignore-read-groups --reference haploid.fasta vcf/chr1.vcf.gz poly.hifi.addrg.sorted.bam 

## UL only
whatshap polyphase -p 4 -o poly.whatshap.chr1.vcf.gz --threads 64 --verify-genotypes --ignore-read-groups --reference haploid.fasta vcf/chr1.vcf.gz poly.ont.addrg.sorted.bam
```

**Comment by schrins** [comment]
I see two issues here:

1. There is no mechanism in Whatshap to deal with different types of reads simultaneously. Internally, all input reads are treated the same, while the ultra long reads have a much higher error rate than the hifi reads. But I don't think this will have a large impact here, in contrast to combinations like hifi reads and 10X reads, where the different sparsity could cause problems.

2. The real problem seems to be, that the blocks with just the hifi reads are already very short. In default settings whatshap cuts the phasing blocks whenever it is insecure about the continuity of haplotypes, which can quickly lead to super-short blocks, whenever something is not right. This means that ultra long reads do not add continuity here, because the blocks are cut for different reasons.

For the latter, there might be two reasons: I personally have not the best experiences with the `--verify-genotypes` flag, so I would not use it. It tends to leave out many variants because of some deviation between reported genotype and counted alleles among the reads and thus lead to more block cuts. Another potential problem could be the realignment for which the reference genome is used. In theory it should lead to better results, but I have also experienced opposite behaviour in some cases. Especially with plant genomes (that is what your data looks like considering the high SNP density) the hifi reads are probably cleaner than the reference genome itself. The realignment was originally designed for diploid human phasing and is - in my opinion - not always recommended for plant data.

In short: Rerun the hifi reads without `--verify-genotypes` and without the `--reference` parameter and see whether your block lengths get better.

**Comment by baozg** [comment]
Thanks for promptly response. HiFi data and reference are different plant accessions. 
I have try three parameters for whatshap. In the whatshap polyphase paper, it use the corrected Nanopore data and ran the command with `--verfify-genotypes`. In the three sets with HiFi only, it seems only use the `--verify-genotypes` have the best performance.

1. with `--reference` , remove `--verify-genotypes`
```bash
---------------- Chromosome chr01 ----------------
     Variants in VCF:  2439041
        Heterozygous:  2072117 ( 1468840 SNVs)
              Phased:  1407663 ( 1407663 SNVs)
            Unphased:   633545 (not considered below)
          Singletons:    30909 (not considered below)
              Blocks:    57808

Block sizes (no. of variants)
   Median block size:        8.00 variants
  Average block size:       24.35 variants
       Largest block:     1100    variants
      Smallest block:        2    variants

Block lengths (basepairs)
      Sum of lengths: 63864536    bp
 Median block length:      277.00 bp
Average block length:     1104.77 bp
       Longest block:    71142    bp
      Shortest block:        1    bp
           Block N50:      nan    bp
```
2. with `--verify-genotypes` , remove `--reference` (✅)
```bash
---------------- Chromosome chr01 ----------------
     Variants in VCF:  2439041
        Heterozygous:  2437447 ( 1834170 SNVs)
              Phased:   529806 (  529806 SNVs)
            Unphased:  1905867 (not considered below)
          Singletons:     1774 (not considered below)
              Blocks:    10410

Block sizes (no. of variants)
   Median block size:       14.00 variants
  Average block size:       50.89 variants
       Largest block:     1344    variants
      Smallest block:        2    variants

Block lengths (basepairs)
      Sum of lengths: 62653691    bp
 Median block length:     1336.00 bp
Average block length:     6018.61 bp
       Longest block:   170039    bp
      Shortest block:        1    bp
           Block N50:      nan    bp
```
3. remove `--reference` and `--verifify`

```bash

---------------- Chromosome chr01 ----------------
     Variants in VCF:  2439041
        Heterozygous:  2073939 ( 1470662 SNVs)
              Phased:  1411223 ( 1411223 SNVs)
            Unphased:   633172 (not considered below)
          Singletons:    29544 (not considered below)
              Blocks:    51306

Block sizes (no. of variants)
   Median block size:        8.00 variants
  Average block size:       27.51 variants
       Largest block:     1652    variants
      Smallest block:        2    variants

Block lengths (basepairs)
      Sum of lengths: 64948385    bp
 Median block length:      259.00 bp
Average block length:     1265.90 bp
       Longest block:   136180    bp
      Shortest block:        1    bp
           Block N50:      nan    bp

```
